src/macaron/vsa/vsa.py

```diff
@@ -0,0 +1,251 @@
+
+    Parameters
+    ----------
+    policy_result : dict
+        The result of the policy engine, including two relations:
+        ``component_violates_policy``, and ``component_satisfies_policy``.
+
+    Returns
+    -------
+    tuple[str, VerificationResult] | None
+        A pair of PURL and verification result of the only software component that
+        the policy applies to, or ``None`` according to the aforementioned conditions.
+    """
+    component_violates_policy_facts = policy_result.get("component_violates_policy", [])
+    component_satisfies_policy_facts = policy_result.get("component_satisfies_policy", [])
+
+    # key: PURL; value: result with the highest component id
+    component_results: dict[str, tuple[int, VerificationResult]] = {}
+
+    for component_id_string, purl, _ in component_violates_policy_facts:
+        component_id = int(component_id_string)
+        if purl not in component_results:
+            component_results[purl] = (component_id, VerificationResult.FAILED)
+        else:
+            current_component_id, _ = component_results[purl]
+            if component_id > current_component_id:
+                component_results[purl] = (component_id, VerificationResult.FAILED)
+    for component_id_string, purl, _ in component_satisfies_policy_facts:
+        component_id = int(component_id_string)
+        if purl not in component_results:
+            component_results[purl] = (component_id, VerificationResult.PASSED)
+        else:
+            current_component_id, _ = component_results[purl]
+            if component_id > current_component_id:
+                component_results[purl] = (component_id, VerificationResult.PASSED)
+
+    if len(component_results) != 1:
+        return None
+
+    subject_purl = next(iter(component_results.keys()))
+    _, verification_result = component_results[subject_purl]
+
+    return subject_purl, verification_result
+
+
+def generate_vsa(policy_content: str, policy_result: dict) -> Vsa | None:
+    """Generate a VSA, if appropriate, based on the result of the policy engine.
+
+    Parameters
+    ----------
+    policy_content : str
+        The Souffle policy code defining the policy.
+    policy_result : dict
+        The result of the policy engine.
+
+    Returns
+    -------
+    Vsa | None
+        The VSA, or ``None`` if generating a VSA is not appropriate according
+        to the policy engine result.
+    """
+    subject_verification_result = get_subject_verification_result(policy_result)
+
+    if subject_verification_result is None:
+        return None
+
+    subject_purl, verification_result = subject_verification_result
+
+    payload = create_vsa_statement(
+        subject_purl=subject_purl,
+        policy_content=policy_content,
+        verification_result=verification_result,
+    )
+
+    return Vsa(
+        payloadType="application/vnd.in-toto+json",
+        payload=base64.b64encode(json.dumps(payload).encode()).decode("ascii"),
```

Comment:
Looks like the errors raised by `base64.b64encode()`, and  `json.dumps` are not caught here?

Reply:
Hmm, I should probably extract this out to an encode function and do all error handling there.

Reply:
Where do you think I should put the encode function? For simplicity, I would just put it in the same file, before we decide what to do with the `intoto` package, currently under the `slsa_analyzer` package.

Reply:
I think at this point we don't need a function and hoisting the code and handling errors should be fine.

Reply:
See d5c6a7935e75d7d5cc01e34942a2d2f645fd6830.